hedera-node/src/main/java/com/hedera/services/grpc/marshalling/ImpliedTransfersMarshal.java

```diff
@@ -0,0 +1,89 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.hedera.services.context.properties.GlobalDynamicProperties;
+import com.hedera.services.ledger.BalanceChange;
+import com.hedera.services.ledger.PureTransferSemanticChecks;
+import com.hedera.services.store.models.Id;
+import com.hederahashgraph.api.proto.java.CryptoTransferTransactionBody;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hedera.services.ledger.BalanceChange.hbarAdjust;
+import static com.hedera.services.ledger.BalanceChange.tokenAdjust;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.OK;
+
+/**
+ * Contains the logic to translate from a gRPC CryptoTransfer operation
+ * to a validated list of balance changes, both ℏ and token unit.
+ *
+ * Once custom fees are implemented for HIP-18, this translation will
+ * become somewhat more complicated, since it will need to analyze the
+ * token transfers for any custom fee payments that need to be made.
+ *
+ * (C.f. https://github.com/hashgraph/hedera-services/issues/1587)
+ */
+public class ImpliedTransfersMarshal {
+	private final GlobalDynamicProperties dynamicProperties;
+	private final PureTransferSemanticChecks transferSemanticChecks;
+
+	public ImpliedTransfersMarshal(
+			GlobalDynamicProperties dynamicProperties,
+			PureTransferSemanticChecks transferSemanticChecks
+	) {
+		this.dynamicProperties = dynamicProperties;
+		this.transferSemanticChecks = transferSemanticChecks;
+	}
+
+	public ImpliedTransfers marshalFromGrpc(CryptoTransferTransactionBody op) {
+		final var maxTokenAdjusts = dynamicProperties.maxTokenTransferListSize();
+		final var maxHbarAdjusts = dynamicProperties.maxTransferListSize();
+
+		final var validity = transferSemanticChecks.fullPureValidation(
+				maxHbarAdjusts, maxTokenAdjusts, op.getTransfers(), op.getTokenTransfersList());
+		if (validity != OK) {
+			return ImpliedTransfers.invalid(maxHbarAdjusts, maxTokenAdjusts, validity);
+		}
+
+		final List<BalanceChange> changes = new ArrayList<>();
+		for (var aa : op.getTransfers().getAccountAmountsList()) {
+			final var grpcAccountId = aa.getAccountID();
+			final var hbarChange = hbarAdjust(Id.fromGrpcAccount(grpcAccountId), aa.getAmount());
+			hbarChange.setExplicitAccountId(grpcAccountId);
+			changes.add(hbarChange);
+		}
+		for (var scopedTransfers : op.getTokenTransfersList()) {
+			final var grpcTokenId = scopedTransfers.getToken();
+			final var scopingToken = Id.fromGrpcToken(grpcTokenId);
+			for (var aa : scopedTransfers.getTransfersList()) {
+				final var grpcAccountId = aa.getAccountID();
+				final var tokenChange = tokenAdjust(scopingToken, Id.fromGrpcAccount(grpcAccountId), aa.getAmount());
+				tokenChange.setExplicitTokenId(grpcTokenId);
+				tokenChange.setExplicitAccountId(grpcAccountId);
```

Comment:
These `setExplicitAccountId` and `setExplicitTokenId` are only used here to avoid multiple conversion from model Id and gRPC account/token Id. I think it's better to move the conversion logic into the `BalanceChange` class and remove those functions altogether. Doing that will simplify the code and avoid exposing unnecessary setters.